(FM-7222) Error out instead of attemping to delete default Cisco VLANs

willmeek opened the pull request.

shermdog commented on lib/puppet/provider/network_vlan/ios.rb:
```diff
@@ -76,8 +76,17 @@ def create(context, name, should)
 
+    def self.check_vlan_can_be_deleted(name)
+      if ['1', '1002', '1003', '1004', '1005'].include?(name)
+        warn "Vlan #{name} is a Cisco default vlan and may not be deleted."
```
Total nit - but the message should be VLAN - right now it's mixed case of Vlan and vlan.

willmeek replied:
Good call, worth getting case right, I've changed these to VLAN

shermdog commented:
I was able to test this.  We should throw errors vs warnings on modifications as the resource implies it's been successfully deleted although it has not.  This breaks idempotency.

We are also unable to modify the name or shutdown status of the reserved VLANs, so we should limit them at create and update as well.

Not sure why, but we're not catching the device error in those cases

```
cat-2960g(config)#vlan 1
cat-2960g(config-vlan)#name test
%Default VLAN 1 may not have its name changed.
cat-2960g(config-vlan)#no shutdown 
%Command is only allowed on VLAN 2..1001.
cat-2960g(config)#vlan 1002
cat-2960g(config-vlan)#name test
%Default VLAN 1002 may not have its name changed.
cat-2960g(config-vlan)#no shutdown
%Command is only allowed on VLAN 2..1001.

willmeek commented:
@shermdog Thanks for getting testing this.

I've moved the validation to be performed on all create/update/delete actions.

The message is now raised as an error, rather than a warning.

And 'belt and braces' I've added those error messages to be checked on send_command

shermdog commented on lib/puppet/util/network_device/cisco_ios/device.rb:
```diff
@@ -36,18 +36,20 @@ def send_command(connection_to_use, options, debug = false)
-      access_denied = commands['default']['access_denied']
-      command_authorization_failed = commands['default']['command_authorization_failed']
-      error_in_authentication = commands['default']['error_in_authentication']
+      access_denied = Regexp.new(%r{#{commands['default']['access_denied']}})
```
I believe these Regexp.new changes are redundant as %r{} is generating a regex pattern already

```ruby
irb(main):018:0> %r{^.*Command is only allowed on VLAN.*$} == Regexp.new(%r{^.*Command is only allowed on VLAN.*$})
=> true
irb(main):031:0> %r{^.*Command is only allowed on VLAN.*$}.class
=> Regexp
irb(main):032:0> Regexp.new(%r{^.*Command is only allowed on VLAN.*$}).class
=> Regexp
```

shermdog commented:
Updates do not seem to be getting caught by validation.

It appears to be the passing of name vs id

```
Notice: /Stage[main]/Main/Node[cat-2960]/Network_vlan[1]/vlan_name: vlan_name changed 'default' to 'default1'
Error: network_vlan[1]: Updating: Failed after 0.173767 seconds: 'name default1
%Default VLAN 1 may not have its name changed.
cat-2960g(config-vlan)#' Error sending 'name default1'
Error: /Stage[main]/Main/Node[cat-2960]/Network_vlan[1]: Could not evaluate: Execution encountered an error
```
```
[1] pry(Puppet::Provider::NetworkVlan::NetworkVlan)> instance[:name]
[2] pry(Puppet::Provider::NetworkVlan::NetworkVlan)> instance
=> {:id=>"1", :ensure=>"present", :vlan_name=>"default1", :shutdown=>false}
```

willmeek commented:
@shermdog Thanks, I've changed the bad logic to match the ID rather than vlan_name. (I'd originally got confused as the RSAPI functions treat the ID as the name)

Added appropriate acceptance test.

shermdog commented:
@willmeek got back around to testing this.  Looks like we are no longer catching deletes `ensure => 'absent'`

Appears to be based on the information we're passing in to `commands_from_instance`.  In the update case we are passing in `should`, which contains `id`.  Delete/Create passes in a hash that only has `name`.

willmeek commented:
Conflicts.

Revisit this as part of the network_vlan phase 2

DavidS commented:
Rebased, resolved conflicts and updated to transports

Thomas-Franklin commented on spec/unit/puppet/provider/network_vlan/cisco_ios_spec.rb:
```diff
@@ -18,5 +30,30 @@ def self.load_test_data
+
+  context 'Vlan non-defaults do not throw error tests:' do
+    ['2', '1001', '1006'].each do |vlan_name|
+      ['absent', 'create', 'present'].each do |_action|
```
There is a loop here but inside the expect it is always `absent`?

DavidS replied:
removed the loop. seems to be a copy/paste remnant from the test above

Thomas-Franklin commented on spec/unit/puppet/provider/network_vlan/cisco_ios_spec.rb:
```diff
@@ -8,7 +8,19 @@ def self.load_test_data
     PuppetX::CiscoIOS::Utility.load_yaml(File.expand_path(__dir__) + '/test_data.yaml', false)
   end
 
+  def self.action_name(action)
```
This method does not appear to be called

DavidS replied:
deleted

DavidS commented:
ad-hoc failed only on the missing `allow_errors` change from https://github.com/puppetlabs/cisco_ios/compare/fc19c81738001628093f43ac9935d54b0cdca561..b6426c6f74e9440bef5b90434840ed831e80307b

Thomas-Franklin commented on spec/acceptance/network_vlan_spec.rb:
```diff
@@ -1,55 +1,81 @@
+      }
+      EOS
+      make_site_pp(pp)
+      expect(run_device(allow_changes: true, allow_errors: true).output).to include('VLAN 1 is a Cisco default VLAN and may not be updated')
```
Receiving:

```
undefined method `output' for #<String:0x00007fcf7245f148>
```

For these two error test cases.

DavidS replied:
fixed.

DavidS commented:
This now passes acceptance tests on at least one catalyst:

```
david@davids:~/git/cisco_ios$ BEAKER_provision=no BEAKER_destroy=no PUPPET_INSTALL_TYPE=pe BEAKER_set=vmpooler bundle exec rspec spec/acceptance/network_vlan_spec.rb --fail-fast

network_vlan
  handling regular VLANs
    creates a network VLAN
    edit a network VLAN
    delete a network VLAN
  handling default VLANs
    errors on update
    errors on delete

Finished in 1 minute 29.08 seconds (files took 0.1792 seconds to load)
5 examples, 0 failures

david@davids:~/git/cisco_ios$
```
